### ISSUE TITLE
Registering mammograms

### ISSUE THREAD

**Issue by ali-yar** [voxelmorph/voxelmorph#51]
Hi @adalca, I am trying to register mammogram images (2D) using voxelmorph CVPR version.

- Images are resized (512x448), normalized (0 - 1) and affinely aligned.
- Registration (training and testing) happens between 2 images of the same breast taken at different time.

![image](https://user-images.githubusercontent.com/32914277/57968261-3a1c3e80-7968-11e9-9792-3a477c514938.png)

1- The breast tissues in the warped/moved image are not smooth (look like flakes). Is it normal or can something be done to improve it?

2- Some calcifications that are visible in the moving image (and have their matches in the fixed image) become invisible after registration. Normal?

3- I would like to measure the target registration error. The landmarks would be the calcifications. With Elastix tool, after registration, a transform parameters file is generated which I use to transform the landmark coordinates and obtain the new coordinates and measure the distance with those landmarks in the fixed image. How can I achieve something similar using voxelmorph?

**Comment by adalca** [comment]
hi @ali-yar 

1 -- I agree, it looks like the field is not regularized enough. Since it's a very different problem from the brain MRIs we worked on, this is not surprising. There's a couple of things you can do, the simplest being to increase the regularization penalty. The second would be to make the field diffeomorphic, which would simply involve adding a call to the `VecInt` layer after the `flow` layer is created in the cvpr network.

2. I would assume this is related to #1 as the field is probably trying hard to fit other features and sometimes will kill some aspects by squishing them too much. It might help to visualize the deformation fields themselves (as images, to start)

3. The default model will give you the displacement field in the target (fixed) space as the second output. This means that you can take the landmarks in the fixed (second) space and warp them to the moving (first) image easily -- (a) you can simply use scipy's `interpn` to get the displacement values *at* the locations of your "fixed" landmarks, (b) add this displacement do their  location, (c) compare with the landmarks in the 'moving' space. 

We have tensorflow code that warps points with a given flow that I am happy to share (it would be more efficient than numpy and you could do it directly when running a network), but I warn you that this hasn't been tested thoroughly and we're only using it internally for now. Or you can build your own tensorflow version using the `interpn` in https://github.com/adalca/neuron/blob/master/neuron/utils.py.

**Comment by ali-yar** [comment]
Thank you for the feedback.

> There's` a couple of things you can do, the simplest being to increase the regularization penalty.

Do you mean the value of `reg_param` as in below?
`mg_model.compile(optimizer=Adam(lr=lr),  loss=[data_loss, losses.Grad('l2').loss],
                     loss_weights=[1.0, reg_param]),`

Also, are the 2 losses (data_loss and l2) working together or independently during optimization?  And does the weight value for one affect the other?

Btw, in the results shown before, the data_loss is 'mse', I have now tried with 'ncc' with gives better visual smoothness but lower mutual information than with mse.

![image](https://user-images.githubusercontent.com/32914277/57973775-52af4780-79ae-11e9-93d4-dfa1bb40cbef.png)

**Comment by adalca** [comment]
Yes, I meant increasing `reg_param`. The two losses are together, that is `loss = data_loss + lambda_param * l2_grad`. The reg_param affects the balance between them.

Note that running this with CC will require a different reg_param (as I think we say in the comments and readme). For brains, we found that reg_param 0.01-0.02 works best, and for CC reg_param of 1-1.5 works best. 

What are you using mutual information for, here? If you consider mutual information a better metric, then you might want to optimize this instead of NCC, although since the tissue differences seem monotonic, I suspect NCC would be the right choice between those two?

Either way, it sounds to me like NCC might do what you want, and you might want to evaluate with the landmarks now?

**Comment by ali-yar** [comment]
I used mutual information as an index to compare the pair of images before and after registration. I will keep NCC.

Sorry to ask you a couple more questions as I am new to this field and I am trying to clarify some concepts. This will help understanding your explanations of points 2 and 3 above.

When registering 2 images (predict), the model returns 2 objects. The second is the registration / deformation field (I assume they are interchangeable terms). It has 2 channels. One shows the displacement in x-direction and the other in y-direction. If I display them separately I get this:

![image](https://user-images.githubusercontent.com/32914277/57976205-8f913380-79da-11e9-831b-7f0266a10f11.png)

If I want to display them as one image? I tried stacking a zero-image as the third channel and got this

![image](https://user-images.githubusercontent.com/32914277/57977376-7b0f6400-79f7-11e9-8677-c46150364df4.png)

**Comment by adalca** [comment]
yeah, all of that is fine. You can also use the `flow` function in `neuron.plot`, which will show you a figure with arrows indicating the actual field.

From the top images, though, you can clearly see that the field is not smooth but it's overfitting to local features -- you can see these small areas everywhere of sudden deformation changes (black to white), etc. So I would definately recommend a higher smoothness penalty for this task.

**Comment by ali-yar** [comment]
> We have tensorflow code that warps points with a given flow that I am happy to share 

I would appreciate if you can share the code for this.

**Comment by adalca** [comment]
Here's a quick piece of code. Note that we haven't tested this thoroughly.
`nrn_utils` refers to `neuron.utils`
note that the displacement field that moves image A to image B will be "in the space of B" -- that is, `trf(p)` tells you "how to move data from A to get to location `p` in B". Therefore, that same displacement field will warp *landmarks* in B to A easily (that is, for any landmark `L(p)`, it can easily find the appropriate `trf(L(p))` via interpolation, as we do below.

```
def point_spatial_transformer(surface_points, trf, single=False):
    """
    surface_points is a N x D or a N x (D+1) Tensor
    trf is a *volshape x D Tensor
    """

    surface_pts_D = surface_points.get_shape().as_list()[-1]
    trf_D = trf.get_shape().as_list()[-1]
    assert surface_pts_D in [trf_D, trf_D + 1]

    if surface_pts_D == trf_D + 1:
        li_surface_pts = K.expand_dims(surface_points[..., -1], -1)
        surface_points = surface_points[..., :-1]

    # just need to interpolate.
    # at each location determined by surface point, figure out the trf...
    # Note: if surface_points are on the grid, gather_nd should work as well
    fn = lambda x: nrn_utils.interpn(x[0], x[1])
    diff = tf.map_fn(fn, [trf, surface_points], dtype=tf.float32)
    ret = surface_points + diff

    if surface_pts_D == trf_D + 1:
        ret = tf.concat((ret, li_surface_pts), -1)

    return ret
```

**Comment by ali-yar** [comment]
Thank you for this.

**Comment by John1231983** [comment]
@adalca : We often have mutual information (MI) loss and NCC loss for registration. Could you tell me when we will MI loss and when we will use NCC loss? Thanks

**Comment by adalca** [comment]
@John1231983 this goes much deeper than voxelmorph, but basically we use NCC when the relationship of anatomical intensities between the two images is monotonic. So if structure A is "darker" than structure B in scan 1, it's also darker in scan 2. For example, you might use this in T1 to T1 registration when you have a lot of noise of you haven't normalized the T1s. We use MSE when these intensities are the *same* modulo noise (e.g. in cleanly normalized T1 to T1). Mutual information of intensities has no such requirement, and you could use it between modalities (E.g. T1 to T2).

**Comment by John1231983** [comment]
Very clear. So if we use a model to learn the appearance changing (likes your cvpr 2019 group), then use registration with mse loss, do you think it will close to the method which apply ncc on original image ( without appearance adding)?

> input ->appearance -> reg with mse

>input -> reg with ncc

**Comment by adalca** [comment]
If the data is very clean, then yeah those would probably give you similar results. But ncc is generally more robust (as many papers have shown), because it computes a spatially local statistic rather than voxel-wise metric, so that means it will deal with noise and small variations "better".
